src/rootfstreemanager.cc

```diff
@@ -24,11 +24,27 @@ bool RootfsTreeManager::fetchTarget(const Uptane::Target &target, Uptane::Fetche
 }
 
 std::string RootfsTreeManager::getServerUrl(RequestHeaders &auth_header) {
-  (void)auth_header;
+  const auto resp = http_client_->post(gateway_url_ + "/download-urls", Json::Value::null);
+
+  if (!resp.isOk()) {
+    LOG_WARNING << "Failed to obtain download URLs from Gateway, fallback to dowload via gateway/proxy server: "
+                << resp.getStatusStr();
+    return "";
+  }
+
+  const auto respJson = resp.getJson();
+  for (Json::ValueConstIterator it = respJson.begin(); it != respJson.end(); it++) {
+    // just try the first URL for now
+    auth_header["Authorization"] = std::string("Bearer ") + (*it)["access_token"].asString();
+    return (*it)["download_url"].asString();
+  }
+
   return "";
 }
 
 void RootfsTreeManager::setRemote(const std::string &name, const std::string &url) {
   OSTree::Repo repo{sysroot_->path() + "/ostree/repo"};
+
+  LOG_DEBUG << "Setting remote to " << url;
```

Comment:
i bet we'll wind up wanting this to be LOG_INFO for support

Reply:
updated

Reply:
i think things look pretty good. lets try and get this staged so we can all start playing with it.